pkg/controllers/trace_controller.go

```diff
@@ -88,8 +121,21 @@ func (r *TraceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 	// Lookup factory
 	factory, ok := r.TraceFactories[trace.Spec.Gadget]
 	if !ok {
-		log.Errorf("Unknown gadget %q", trace.Spec.Gadget)
-		return ctrl.Result{}, nil
+		return ctrl.Result{}, setTraceOpError(ctx, r.Client, req.NamespacedName.String(),
+			trace, fmt.Sprintf("Unknown gadget %q", trace.Spec.Gadget))
+	}
+
+	// Check trace specs before adding the finalizer and registering the trace
+	if trace.Spec.RunMode != "Manual" {
+		return ctrl.Result{}, setTraceOpError(ctx, r.Client, req.NamespacedName.String(),
+			trace, fmt.Sprintf("Unsupported RunMode %q for gadget %q",
+				trace.Spec.RunMode, trace.Spec.Gadget))
+	}
+	outputModes := factory.OutputModesSupported()
+	if _, ok := outputModes[trace.Spec.OutputMode]; !ok {
+		return ctrl.Result{}, setTraceOpError(ctx, r.Client, req.NamespacedName.String(),
+			trace, fmt.Sprintf("Unsupported OutputMode %q for gadget %q",
+				trace.Spec.OutputMode, trace.Spec.Gadget))
```

Comment:
Considering this scenario:
1. User creates a valid Trace resource
2. The controller will add the finalizer
3. User edits the resource with `kubectl edit` to switch the OutputMode to an unsupported mode
4. The controller will write the error message in OperationError
5. User deletes the resource
6. The controller will never complete the deletion.

Moving this before the deletion code means that the controller might never finish to delete the trace.

Reply:
Oh, you're right. I missed considering the possibility of editing the trace with `kubectl` or somewhere outside the IG.

I agree it should not be moved before the deletion code. But the same reasoning goes also for the factory lookup in the current code thus I will move all of them right after the deletion code. Do you agree?

Reply:
Also, the Node is suffering the same issue so I will create an issue to work on it in a future PR.

Reply:
Issue: https://github.com/kinvolk/inspektor-gadget/issues/329